drivers/mtd/mtd_config_fs.c

```diff
@@ -0,0 +1,2126 @@
+ *   fs    - Pointer to file system.
+ *   pdata - Pointer to data buffer.
+ *
+ * Returned Value:
+ *   Number of bytes written. On success, it will be equal to the
+ *   number of bytes requested to be written. On error returns -ERRNO code.
+ *
+ ****************************************************************************/
+
+static ssize_t nvs_write(FAR struct nvs_fs *fs,
+                         FAR struct config_data_s *pdata)
+{
+  int rc;
+  int gc_count;
+  size_t data_size;
+  size_t key_size;
+  struct nvs_ate wlk_ate;
+  uint32_t wlk_addr;
+  uint32_t rd_addr;
+  uint32_t hist_addr;
+  uint16_t required_space = 0;
+  bool prev_found = false;
+  uint32_t hash_id;
+  uint16_t block_to_write_befor_gc;
+  uint32_t second_latest_ate_addr;
+
+#ifdef CONFIG_MTD_CONFIG_NAMED
+  FAR const uint8_t *key;
+#else
+  uint8_t key[sizeof(pdata->id) + sizeof(pdata->instance)];
+#endif
+
+#ifdef CONFIG_MTD_CONFIG_NAMED
+  key = (FAR const uint8_t *)pdata->name;
+  key_size = strlen(pdata->name) + 1;
+#else
+  memcpy(key, &pdata->id, sizeof(pdata->id));
+  memcpy(key + sizeof(pdata->id), &pdata->instance, sizeof(pdata->instance));
+  key_size = sizeof(pdata->id) + sizeof(pdata->instance);
+#endif
+
+  /* Data now contains input data and input key, input key first. */
+
+  data_size = key_size + pdata->len;
+
+  /* The maximum data size is block size - 3 ate
+   * where: 1 ate for data, 1 ate for block close, 1 ate for gc done.
+   */
+
+  finfo("key_size=%zu, len=%zu\n", key_size, pdata->len);
+
+  if ((data_size > (fs->geo.erasesize - 3 * sizeof(struct nvs_ate))) ||
+      ((pdata->len > 0) && (pdata->configdata == NULL)))
+    {
+      return -EINVAL;
+    }
+
+  /* Calc hash id of key. */
+
+  hash_id = nvs_fnv_hash(key, key_size) % 0xfffffffd + 1;
+
+  /* Find latest entry with same id. */
+
+  wlk_addr = fs->ate_wra;
+
+  while (1)
+    {
+      rd_addr = wlk_addr;
+      hist_addr = wlk_addr;
+      rc = nvs_prev_ate(fs, &wlk_addr, &wlk_ate);
+      if (rc)
+        {
+          return rc;
+        }
+
+      if ((wlk_ate.id == hash_id) && (nvs_ate_valid(fs, &wlk_ate)))
+        {
+          if ((wlk_ate.key_len == key_size)
+              && (!nvs_flash_block_cmp(fs,
+              (rd_addr & ADDR_BLOCK_MASK) + wlk_ate.offset, key, key_size)))
+            {
+              prev_found = true;
+              break;
+            }
+          else
+            {
+              fwarn("hash conflict\n");
+            }
+        }
+
+      if (wlk_addr == fs->ate_wra)
+        {
+          break;
+        }
+    }
+
+  if (prev_found)
+    {
+      finfo("Previous found\n");
+
+      /* Previous entry found. */
+
+      rd_addr &= ADDR_BLOCK_MASK;
+
+      if (pdata->len == 0)
+        {
+          /* If prev ate is expired, it is deleted. */
+
+          if (wlk_ate.expired != fs->erasestate)
+            {
+              /* Skip delete entry as it is already the
+               * last one.
+               */
+
+              return 0;
+            }
```

Comment:
why not directly expire here?

Reply:
The case is if we want to delete one , and prev one is already deleted, so nothing needs to be done